### PR TITLE
PRT - fix e2e race condition and invalid behavior

### PR DESCRIPTION
**Pull request by ranlavanet**

### CODE REVIEW

File: x/pairing/types/QualityOfServiceReport.go [hunk 1]
```diff
@@ -15,3 +15,12 @@ func (qos *QualityOfServiceReport) ComputeQoS() (sdk.Dec, error) {
...
 	return qos.Availability.Mul(qos.Sync).Mul(qos.Latency).ApproxRoot(3)
 }
+
+func (qos *QualityOfServiceReport) ComputeQoSExcellence() (sdk.Dec, error) {
+	if qos.Availability.LT(sdk.ZeroDec()) ||
+		qos.Latency.LT(sdk.ZeroDec()) ||
+		qos.Sync.LT(sdk.ZeroDec()) {
+		return sdk.ZeroDec(), fmt.Errorf("QoS excellence scores is below 0")
+	}
+	return qos.Availability.Quo(qos.Sync).Quo(qos.Latency).ApproxRoot(3)
```
Comment (omerlavanet) [inline x/pairing/types/QualityOfServiceReport.go:25]: oops, zero division

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Review by omerlavanet — DISMISSED** [review]

**Review by omerlavanet — DISMISSED** [review]

**Review by omerlavanet — APPROVED** [review]

File: x/pairing/types/QualityOfServiceReport.go [hunk 2]
```diff
@@ -15,3 +15,12 @@ func (qos *QualityOfServiceReport) ComputeQoS() (sdk.Dec, error) {
...
 	return qos.Availability.Mul(qos.Sync).Mul(qos.Latency).ApproxRoot(3)
 }
+
+func (qos *QualityOfServiceReport) ComputeQoSExcellence() (sdk.Dec, error) {
+	if qos.Availability.LTE(sdk.ZeroDec()) ||
+		qos.Latency.LTE(sdk.ZeroDec()) ||
+		qos.Sync.LTE(sdk.ZeroDec()) {
+		return sdk.ZeroDec(), fmt.Errorf("QoS excellence scores is below 0")
+	}
+	return qos.Availability.Quo(qos.Sync).Quo(qos.Latency).ApproxRoot(3)
```
Comment (Yaroms) [inline x/pairing/types/QualityOfServiceReport.go:25]: why not use the objects method instead?

Comment (Yaroms) [reply x/pairing/types/QualityOfServiceReport.go:25]: ohh different, nvm

File: scripts/init_e2e.sh [hunk 3]
```diff
@@ -62,14 +56,22 @@ lavad tx subscription buy "EmergencyModePlan" -y --from user5 --gas-adjustment "
 # Test plan upgrade
 echo ---- Subscription plan upgrade ----
 wait_next_block
-lavad tx subscription buy "DefaultPlan" -y --from user1 --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
+# test we have the plan active.
+current_plan=$(lavad q subscription current $(lavad keys show user1 -a)) # store current plan in a different variable to print in case of an error
+echo "current plan before upgrade: $current_plan"
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply scripts/init_e2e.sh:61]: just printing it. otherwise we dont have the info properly before it was undefined so i added it to debug

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**